service/storage_proxy.cc

```diff
@@ -5317,112 +5303,98 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
     return _hints_manager.get_host_filter();
 }
 
-future<> storage_proxy::create_hint_queue_sync_point(utils::UUID sync_point_id, std::vector<gms::inet_address> endpoints, clock_type::time_point deadline) {
-    return container().invoke_on(0, [sync_point_id, endpoints = std::move(endpoints), deadline] (storage_proxy& sp) mutable {
-        auto [it, was_inserted] = sp._hint_queue_checkpoints.emplace(sync_point_id);
-        if (!was_inserted) {
-            return make_exception_future<>(std::runtime_error(format("Hint sync point {} already exists", sync_point_id)));
-        }
-
-        // Waited indirectly by keeping a pointer to the storage_proxy.
-        // When drain_on_shutdown is triggered, hints manager will report an error
-        // and this future will resolve.
-        (void)sp.container().invoke_on_all([endpoints = std::move(endpoints), sync_point_id, deadline] (storage_proxy& sp) {
-            auto wait_for_hints_manager = [&endpoints, sync_point_id, deadline] (db::hints::manager& mgr, const char* mgr_name) {
-                return mgr.wait_until_hints_are_replayed(endpoints, deadline).then_wrapped([mgr_name, sync_point_id] (future<>&& f) {
-                    if (!f.failed()) {
-                        slogger.debug("Hint sync point {} for {} reached", sync_point_id, mgr_name);
-                        f.get();
-                    } else {
-                        slogger.debug("An error occured when waiting for hint sync point {} for {} to resolve: {}", sync_point_id, mgr_name, f.get_exception());
-                    }
-                });
-            };
-
-            return when_all(
-                wait_for_hints_manager(sp._hints_manager, "hints manager"),
-                wait_for_hints_manager(sp._hints_for_views_manager, "hints for view manager")
-            ).discard_result();
-        }).then([&sp, sync_point_id, guard = sp.shared_from_this()] {
-            sp._hint_queue_checkpoints.erase(sync_point_id);
+future<db::hints::sync_point> storage_proxy::create_hint_sync_point(const std::vector<gms::inet_address> target_hosts) const {
+    db::hints::sync_point spoint;
+    spoint.regular_per_shard_rps.resize(smp::count);
+    spoint.mv_per_shard_rps.resize(smp::count);
+    spoint.host_id = co_await db::system_keyspace::get_local_host_id();
+    co_await parallel_for_each(boost::irange<unsigned>(0, smp::count), [this, &target_hosts, &spoint] (unsigned shard) {
```

Comment:
This `parallel_for_each` kind of looks like a manual implementation of `invoke_on_all` - maybe it will be cleaner to use this primitive instead?

Reply:
I couldn't use `invoke_on_all` here because that method is not const, while `this` in current context is const.

Reply:
I suppose we need a const version of invoke_on_all(), but it will be a pain to add all those overloads.